### PR TITLE
[dist_moe] Add the TorchTitan backend adapter

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #4626
* #4625
* __->__ #4624
* #4623
* #4622
* #4621
* #4620
* #4619
* #4618

Summary:
Add a thin TorchTitan adapter for the standalone Dist-MoE package. TorchTitan owns backend selection, W13/W2 model parameters, FSDP/optimizer integration, checkpoint key conversion, process groups, and context lifetime. The package continues to own dispatch, expert compute, combine, memory planning, activation storage, VMM, and kernels.

MXFP8 expert weights specialize TorchTitan's existing `_ShardedFSDPTensor` contract. They share BF16 all-gather, operand preparation/refill/release, SimpleFSDP gradient edges, RAF semantics, and checkpoint handling with `MXFP8Linear`; only grouped Dist-MoE operand construction differs.

Translate the common route-output callback into `DistMoeExpertPostprocess` while preserving sparse-mesh context. The converter rejects specialized routed-expert configs instead of silently dropping model behavior. BF16 and MXFP8 use the same callback contract without changing prepared-weight or in-place WGRAD ownership.

The package selects one of two entries to the same autograd implementation: ordinary eager calls it directly, while Dynamo, FakeTensor, make_fx, and activation-checkpoint tracing use the registered ordered-effect operation. Both entries reach the same saved-state, backward, memory planner, and kernels.

Non-goals:
- Do not copy or modify kernels, routing policy, the memory planner, pipeline schedules, or the generic prepared-weight lifecycle.
- Do not silently support FSDP sharding outside logical dimension 0.
- Do not add model-specific postprocessing or parameter ownership to the adapter.

Dependencies and risks:
- Requires TorchTitan #4203/#4371, #4539, the route-output callback PR below this change, the PyTorch #194114/#196460 FSDP lifecycle, and a released exact `dist-moe` version. Pipeline use additionally requires #4540 and PyTorch #196462-#196465. Effectful AC requires #4559.
- Callback state must obey the annex activation planner. The permanent internal oracle covers saved/recomputed callback inputs and all gradients.

Test Plan:
- `python -m pytest -q tests/unit_tests/cpu/test_moe.py tests/unit_tests/cpu/test_dist_moe.py` (29 passed, including BF16 and MXFP8 callback adaptation)
- `pyrefly check torchtitan/models/common/moe.py tests/unit_tests/cpu/test_moe.py torchtitan/components/dist_moe/backend.py tests/unit_tests/cpu/test_dist_moe.py` (0 errors)
- `buck2 test //genai/msl/dist_moe/fb/tests/parity:test_dist_moe_parity` (3/3 passed: https://www.internalfb.com/intern/testinfra/testrun/3659175089648429)
- Existing two-rank BF16 and MXFP8 FSDP+EP full-CUDA-graph training and final-source MXFP8 device/VMM/GraphTrainer acceptance remain applicable to the unchanged default path.

Pull-Request: https://github.com/pytorch/torchtitan/pull/4541

### CODE REVIEW

**Comment by sanketpurandare** [comment]
Closing an accidental duplicate created because the local stack carried a noncanonical PR trailer. The existing stack PR remains authoritative.
